Fix hardcoded reliance on pip

vpetrigo opened the pull request:
This is an attempt to fix the issue described here juftin/hatch-pip-compile#85.

As it was discussed there, since `hatch` already depends on `uv` there is no need to rely on `pip` install during environment requirements resolution, because it may be not available.

vpetrigo committed:
hatch: cli: application: Switch to uv

Address: juftin/hatch-pip-compile#85

# Conflicts:
#	src/hatch/cli/application.py

vpetrigo committed:
tests: Update asserts after switching to uv

juftin commented on src/hatch/cli/application.py:
```diff
@@ -160,9 +160,9 @@ def ensure_plugin_dependencies(self, dependencies: list[Requirement], *, wait_me
                 return
 
-            pip_command = [sys.executable, '-u', '-m', 'pip']
+            pip_command = [sys.executable, '-u', '-m', 'uv', 'pip']
```
`uv` makes a singular Python import available, `find_uv_bin` - that may be more reliable than `python -m uv ...` 🤔 
https://github.com/astral-sh/uv/blob/54b3a438d03483c6e76f05168bd417657b693e4b/python/uv/_find_uv.py#L8C5-L8C16

This likely works as well though

vpetrigo replied:
Updated!

juftin commented on src/hatch/cli/application.py:
This is great, I was imagining the fix for this would be some complicated piece of work - but this is a really nice, targeted swap of `pip` for `uv pip` only in `ensure_plugin_dependencies`

juftin commented on src/hatch/cli/application.py:
This test is failing for me:
```Dockerfile
FROM fedora:latest

RUN dnf install -y git

COPY --from=ghcr.io/astral-sh/uv:latest /uv /uvx /bin/

RUN uv python install 3.12
RUN uv tool install git+https://github.com/vpetrigo/hatch.git?branch=hardcoded-reliance-on-pip
ENV PATH="/root/.local/bin:${PATH}"

WORKDIR /work

RUN touch /work/pyproject.toml

RUN cat <<EOF > /work/pyproject.toml
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "foo"
version = "0.0.1"

[tool.hatch.env]
requires = [
    "hatch-pip-compile"
]

[tool.hatch.envs.default]
installer = "uv"
EOF

RUN mkdir /work/foo
RUN touch /work/foo/__init__.py

CMD ["hatch", "run", "ls"]
```

```shell
docker build . --tag hatch-pip-compile-test
docker run --rm -it hatch-pip-compile-test
```

juftin replied:
This actually fails now since we've replaced `sys.executable -m pip` with `find_uv_bin` 😢 

Looks like `uv` doesn't know what environment to install into

```
error: No virtual environment found; run `uv venv` to create an environment, or pass `--system` to install into a non-virtual environment
```

vpetrigo replied:
Rolled back for now to `[sys.executable, '-u', '-m', 'uv', 'pip']`. That is interesting that `uv` fails even before the virtual environment is actually created. 🤔

vpetrigo replied:
What may work is the following command:
```python
uv_bin = find_uv_bin()
pip_command = [uv_bin, 'pip']

pip_command.extend(['install', '--directory', str(pathlib.Path(sys.executable).parent.parent)])
```

Not sure if it is more reliable than calling a `uv` module with `venv` Python.

vpetrigo replied:
Pushed those changes for review just in the case that approach with `find_uv_bin` is preferable. 😅

juftin replied:
I was thinking about two options that would work here:

The first option essentially calls `python -m uv pip` which when invoked as a module discovers the Python environment it belongs to 

```py
pip_command = [sys.executable, "-m", "uv", "pip"]
pip_command.extend(["install"])
```

The second option is to call the `find_uv_bin` method, but also passes the current Python interpreter using the [`--python` argument](https://docs.astral.sh/uv/pip/environments/#using-arbitrary-python-environments)

```py
uv_binary = find_uv_bin()
pip_command = [uv_binary, "pip"]
pip_command.extend(["install", "--python", sys.executable])
```

Honestly, I think both of these options do the same thing 🤷 - I'm fine with whatever you and Ofek like

vpetrigo replied:
I updated the implementation for the last option with passing `--python` switch as it is more readable and less confusing than passing venv Python directory with `--directory` switch. Run through the tests in `hatch` and with the container you prepared - all passed on my end.

vpetrigo committed:
hatch: tests: Update uv command search

vpetrigo committed:
hatch: Update to use find_uv_bin()

vpetrigo committed:
hatch: Update the implementation to use --python switch

Switch from --directory to --python switch to simplify `uv pip install` command.

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
misc: Remove unused pathlib imports and switch to consistent string quoting

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
tests: Use consistent double-quoted strings in conftest.py

vpetrigo commented:
@ofek would you please re-approve workflows once again, I fixed formatting errors reported. I am really sorry to bother you. 😔

vpetrigo committed:
tests: Handle commands with "hatchling" in conftest.py mock

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

ofek reviewed (changes requested):
Thanks for your patience here!

vpetrigo committed:
fix: Make find_uv_bin import lazy

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo commented on src/hatch/cli/application.py:
```diff
@@ -5,6 +5,8 @@
 from functools import cached_property
 from typing import TYPE_CHECKING, cast
 
+from uv import find_uv_bin
```
@ofek would you please tell me if any other changes are required, when you have time.😅

vpetrigo replied:
@ofek just a friendly reminder that I am still awaiting for the feedback. 😅

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

cjames23 committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo commented:
Hi @ofek @cjames23,

I wanted to follow up on this pull request. I have addressed a while ago the feedback from the previous review, and the changes should be good to go. I know you all are incredibly busy, but whenever you have some time to take a look, I would really appreciate it.

Thanks again for your time and dedication to the project!

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

vpetrigo committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip

cjames23 commented:
I will work on getting the type issue fixed so we can check the tests here

vpetrigo commented:
@cjames23 please let me know if I can help you with that.

cjames23 committed:
Merge branch 'master' into fix/hardcoded-reliance-on-pip
